File: 367/valid_perfect_square.py
# 367. Valid Perfect Square
# https://leetcode.com/problems/valid-perfect-square/
def main():
  solution = Solution()
  print("=========================================")
  print(solution.isPerfectSquare(1))
  print(True)
  print("=========================================")
  print(solution.isPerfectSquare(4))
  print(True)
  print("=========================================")
  print(solution.isPerfectSquare(9))
  print(True)
  print("=========================================")
  print(solution.isPerfectSquare(16))
  print(True)
  print("=========================================")
  print(solution.isPerfectSquare(14))
  print(False)
  print("=========================================")
  print(solution.isPerfectSquare(2147483647))
  print(False)
  print("=========================================")
  print(solution.isPerfectSquare((1<<63) - 1))
  print(False)
  print("=========================================")
  print(solution.isPerfectSquare(3037000499 ** 2))
  print(True)
  print("=========================================")

class Solution:
  def isPerfectSquare(self, num: int) -> bool:
    start, end = 1, num
    while (start <= end):
        mid = start + (end - start >> 1)
        div = num // mid
        if div == mid and num % mid == 0:
            return True
        elif div > mid:
            start = mid + 1
        else:
            end = mid - 1
    return False

  # def isPerfectSquare(self, num: int) -> bool:
  #   start, end = 1, num
  #   while (start <= end):
  #       mid = (start + end) // 2
  #       square = mid * mid
  #       if square == num:
  #           return True
  #       elif square < num:
  #           start = mid + 1
  #       else:
  #           end = mid - 1
  #   return False

if __name__ == '__main__':
    main()

File: 367/ValidPerfectSquare.kt
// 367. Valid Perfect Square
// https://leetcode.com/problems/valid-perfect-square/
fun main(args: Array<String>) {
    val solution = Solution()
    println("=========================================")
    println(solution.isPerfectSquare(1))
    println(true)
    println("=========================================")
    println(solution.isPerfectSquare(4))
    println(true)
    println("=========================================")
    println(solution.isPerfectSquare(9))
    println(true)
    println("=========================================")
    println(solution.isPerfectSquare(16))
    println(true)
    println("=========================================")
    println(solution.isPerfectSquare(14))
    println(false)
    println("=========================================")
    println(solution.isPerfectSquare(2147483647))
    println(false)
    println("=========================================")
    // println(solution.isPerfectSquare((1 shl 63) - 1))
    // println(false)
    // println("=========================================")
    // println(solution.isPerfectSquare(3037000499 * 3037000499))
    // println(true)
    // println("=========================================")
}

class Solution {
    fun isPerfectSquare(num: Int): Boolean {
        var min = 1
        var max = num
        while (min <= max) {
            val mid = min + (max - min) / 2
            val div = num / mid
            if (div == mid && num % mid == 0) return true
            if (div > mid) {
                min = mid + 1
            } else {
                max = mid - 1
            }
        }
        return false
    }
}
